tests/integration/pilot/reachabiity/mtls_test.go

```diff
@@ -0,0 +1,257 @@
+	// App without sidecar.
+	nakedApp, _ := appsInstance.GetAppOrFail("t", t).(apps.KubeApp)
+
+	testCases := []struct {
+		configFile  string
+		namespace   string
+		connections []connection
+	}{
+		{
+			configFile: "global-mtls-on.yaml",
+			connections: []connection{
+				{
+					from:            aApp,
+					to:              bApp,
+					port:            80,
+					protocol:        apps.AppProtocolHTTP,
+					expectedSuccess: true,
+				},
+				{
+					from:            aApp,
+					to:              headlessApp,
+					port:            80,
+					protocol:        apps.AppProtocolHTTP,
+					expectedSuccess: true,
+				},
+				{
+					from:            nakedApp,
+					to:              bApp,
+					port:            80,
+					protocol:        apps.AppProtocolHTTP,
+					expectedSuccess: false,
+				},
+			},
+		},
+		{
+			configFile: "global-mtls-permissive.yaml",
+			connections: []connection{
+				{
+					from:            aApp,
+					to:              bApp,
+					port:            80,
+					protocol:        apps.AppProtocolHTTP,
+					expectedSuccess: true,
+				},
+				{
+					from:            nakedApp,
+					to:              bApp,
+					port:            80,
+					protocol:        apps.AppProtocolHTTP,
+					expectedSuccess: true,
+				},
+			},
+		},
+		{
+			configFile: "global-mtls-off.yaml",
+			connections: []connection{
+				{
+					from:            aApp,
+					to:              bApp,
+					port:            80,
+					protocol:        apps.AppProtocolHTTP,
+					expectedSuccess: true,
+				},
+				{
+					from:            nakedApp,
+					to:              bApp,
+					port:            80,
+					protocol:        apps.AppProtocolHTTP,
+					expectedSuccess: true,
+				},
+			},
+		},
+		{
+			configFile: "single-port-mtls-on.yaml",
+			namespace:  appsInstance.Namespace().Name(),
+			connections: []connection{
+				{
+					from:            aApp,
+					to:              bApp,
+					port:            80,
+					protocol:        apps.AppProtocolHTTP,
+					expectedSuccess: false,
+				},
+				{
+					from:            nakedApp,
+					to:              bApp,
+					port:            80,
+					protocol:        apps.AppProtocolHTTP,
+					expectedSuccess: false,
+				},
+				{
+					from:            aApp,
+					to:              bApp,
+					port:            90,
+					protocol:        apps.AppProtocolHTTP,
+					expectedSuccess: true,
+				},
+				{
+					from:            nakedApp,
+					to:              bApp,
+					port:            90,
+					protocol:        apps.AppProtocolHTTP,
+					expectedSuccess: true,
+				},
+			},
+		},
+	}
+
+	for _, c := range testCases {
+		namespace := c.namespace
+		if len(namespace) == 0 {
+			namespace = istioCfg.SystemNamespace
+		}
+
+		policy := setupPolicy(t, env, namespace, c.configFile)
+		// Give some time for the policy propagate.
+		// TODO: query pilot or app to know instead of sleep.
+		time.Sleep(time.Second)
+		for _, conn := range c.connections {
+			retry.UntilSuccess(func() error {
```

Comment:
You want to use `UntilSuccessOrFail` so that the test fails if the retry times out.

Reply:
Ah, make sense. Thanks.